File: src/Ddd.Tests/Trainings/Idea/Domain/Tests/TrainingIdeaTest.cs
using Ddd.Tests.Trainings.Tests;
using Ddd.Trainings.Idea.Domain;
using Ddd.Trainings.Proposal.Domain;
using Xunit;

namespace Ddd.Trainings.Idea.Domain.Tests;

public class TrainingIdeaTest : TrainingTests
{
    [Fact]
    public void EditDuration_UpdatesDuration()
    {
        // given
        TrainingIdea idea = TrainingIdeaBuilder.Build();

        // when
        idea.EditDuration(TrainingDurationVO.From(3));

        // then
        Assert.Equal(3, idea.Duration.Days);
    }

    [Fact]
    public void EditDuration_ThrowsException_WhenIdeaIsProposed()
    {
        // given
        TrainingIdea idea = TrainingIdeaBuilder.Proposed(true).Build();

        // when & then
        var exception = Assert.Throws<InvalidOperationException>(() => idea.EditDuration(TrainingDurationVO.From(3)));
        Assert.Equal("Cannot edit duration of proposed idea", exception.Message);
    }

    [Fact]
    public void Propose_CreatesProposal_WhenDurationIsValid()
    {
        // given
        TrainingIdea idea = TrainingIdeaBuilder.Build();

        // when
        TrainingProposal proposal = idea.Propose(FakeReviewerPolicy);

        // then
        Assert.NotNull(proposal);
        Assert.Equal(idea.GetId(), proposal.IdeaId);
        Assert.True(idea.IsProposed);
    }

    [Fact]
    public void Propose_ThrowsException_WhenDurationIsEmpty()
    {
        // given
        TrainingIdea idea = TrainingIdeaBuilder
            .WithDuration(TrainingDurationVO.Empty())
            .Build();

        // when & then
        var exception = Assert.Throws<ArgumentException>(() => idea.Propose(FakeReviewerPolicy));
        Assert.Equal("Duration is empty", exception.Message);
    }
} 

File: src/test/java/com/ddd/trainings/idea/domain/TrainingIdeaTest.java
package com.ddd.trainings.idea.domain;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import com.ddd.trainings.TrainingTests;
import org.junit.jupiter.api.Test;

class TrainingIdeaTest extends TrainingTests {

    @Test
    void editDuration_updatesDuration() {
        // given
        TrainingIdea idea = trainingIdeaBuilder.build();

        // when
        idea.editDuration(TrainingDurationVO.from(3));

        // then
        assertThat(idea.getDuration().getDays()).isEqualTo(3);
    }

    @Test
    void editDuration_throwsException_whenIdeaIsProposed() {
        // given
        TrainingIdea idea = trainingIdeaBuilder.proposed(true).build();

        // when & then
        assertThatThrownBy(() -> idea.editDuration(TrainingDurationVO.from(3)))
            .isInstanceOf(IllegalStateException.class)
            .hasMessage("Cannot edit duration of proposed idea");
    }

    @Test
    void propose_createsProposal_whenDurationIsValid() {
        // given
        TrainingIdea idea = trainingIdeaBuilder.build();

        // when
        // TODO: Implement test that proposes training idea
    }

    @Test
    void propose_throwsException_whenDurationIsEmpty() {
        // given
        TrainingIdea idea = trainingIdeaBuilder
            .withDuration(TrainingDurationVO.empty())
            .build();

        // when & then
        // TODO: verify that idea with invalidation throws exception
    }
} 